flexmeasures_openweathermap/utils/owm.py

```diff
@@ -0,0 +1,204 @@
+    assert (
+        res.status_code == 200
+    ), f"OpenWeatherMap returned status code {res.status_code}: {res.text}"
+    data = res.json()
+    time_of_api_call = as_server_time(
+        datetime.fromtimestamp(data["current"]["dt"], tz=get_timezone())
+    ).replace(second=0, microsecond=0)
+    return time_of_api_call, data["hourly"]
+
+
+def save_forecasts_in_db(
+    api_key: str,
+    locations: List[Tuple[float, float]],
+    data_source: DataSource,
+    max_degree_difference_for_nearest_weather_sensor: int = 2,
+):
+    """Process the response from OpenWeatherMap API into Weather timed values.
+    Collects all forecasts for all locations and all sensors at all locations, then bulk-saves them.
+    """
+    click.echo("[FLEXMEASURES] Getting weather forecasts:")
+    click.echo("[FLEXMEASURES]  Latitude, Longitude")
+    click.echo("[FLEXMEASURES] -----------------------")
+
+    for location in locations:
+        click.echo("[FLEXMEASURES] %s, %s" % location)
+        weather_sensors: Dict[
+            str, Sensor
+        ] = {}  # keep track of the sensors to save lookups
+        db_forecasts: Dict[Sensor, List[TimedBelief]] = {}  # collect beliefs per sensor
+
+        time_of_api_call, forecasts = call_openweatherapi(api_key, location)
+        click.echo(
+            "[FLEXMEASURES] Called OpenWeatherMap API successfully at %s."
+            % time_of_api_call
+        )
+
+        # loop through forecasts, including the one of current hour (horizon 0)
+        for fc in forecasts:
+            fc_datetime = as_server_time(
+                datetime.fromtimestamp(fc["dt"], get_timezone())
+            ).replace(second=0, microsecond=0)
+            fc_horizon = fc_datetime - time_of_api_call
+            click.echo(
+                "[FLEXMEASURES] Processing forecast for %s (horizon: %s) ..."
+                % (fc_datetime, fc_horizon)
+            )
+            for owm_response_label in owm_to_sensor_map:
+                sensor_name = str(owm_to_sensor_map[owm_response_label]["name"])
+                if owm_response_label in fc:
+                    if sensor_name in weather_sensors:
+                        weather_sensor = weather_sensors[sensor_name]
+                    else:
+                        weather_sensor = find_weather_sensor_by_location_or_fail(
+                            location,
+                            max_degree_difference_for_nearest_weather_sensor,
+                            sensor_name=sensor_name,
+                        )
+                        weather_sensors[sensor_name] = weather_sensor
+                    if weather_sensor not in db_forecasts.keys():
+                        db_forecasts[weather_sensor] = []
+
+                    fc_value = fc[owm_response_label]
+                    # the radiation is not available in OWM -> we compute it ourselves
+                    if sensor_name == "radiation":
+                        fc_value = compute_irradiance(
```

Comment:
For irradiance, the data source is not strictly OWM, but rather OWM data in combination with the FlexMeasures model to compute irradiance from cloud coverage. This connects to a larger issue we have been discussing for a while (source lineage, see https://github.com/SeitaBV/timely-beliefs/issues/33), but for now I'd recommend just creating a separate DataSource for irradiance values. That would make it analogous to how we handle CO₂ emission data in flexmeasures-entsoe, i.e. as a derived data source.